File: backend/app/schemas/budget.py
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field

from .departement import DepartementResponse
from .exercice_budgetaire import ExerciceBudgetaireResponse
from .ligne_budgetaire import LigneBudgetaireSimpleResponse
from .projet import ProjetSimpleResponse
from .user import UserSimpleResponse


class BudgetBase(BaseModel):
    reference: Optional[str] = Field(None, min_length=2, max_length=100)
    libelle: str = Field(..., min_length=2, max_length=150)
    description: Optional[str] = None
    devise: str = Field("FC", pattern="^(FC|USD)$", max_length=3)
    montant_total_prevu: Optional[Decimal] = Field(None, ge=0)
    montant_total_realise: Optional[Decimal] = Field(None, ge=0)
    total_recettes_realisees: Optional[Decimal] = Field(None, ge=0)
    total_depenses_realisees: Optional[Decimal] = Field(None, ge=0)
    taux_execution_budgetaire: Optional[Decimal] = Field(None, ge=0)
    ecart_total: Optional[Decimal] = Field(None)
    statut: Optional[str] = Field("brouillon", max_length=50)
    model_config = ConfigDict(extra="forbid")


class BudgetCreate(BudgetBase):
    projet_id: int
    montant_total_prevu: Optional[Decimal] = None
    montant_total_realise: Optional[Decimal] = None
    ecart_total: Optional[Decimal] = None


class BudgetUpdate(BaseModel):
    reference: Optional[str] = Field(None, min_length=2, max_length=100)
    libelle: Optional[str] = Field(None, min_length=2, max_length=150)
    description: Optional[str] = None
    montant_total_prevu: Optional[Decimal] = Field(None, ge=0)
    montant_total_realise: Optional[Decimal] = Field(None, ge=0)
    total_recettes_realisees: Optional[Decimal] = Field(None, ge=0)
    total_depenses_realisees: Optional[Decimal] = Field(None, ge=0)
    taux_execution_budgetaire: Optional[Decimal] = Field(None, ge=0)
    ecart_total: Optional[Decimal] = None
    statut: Optional[str] = Field(None, max_length=50)
    model_config = ConfigDict(extra="forbid")


class BudgetSimpleResponse(BudgetBase):
    id: int
    departement_id: int
    exercice_id: int
    created_by_id: int
    projet_id: Optional[int] = None
    model_config = ConfigDict(from_attributes=True)


class BudgetResponse(BudgetSimpleResponse):
    departement: Optional[DepartementResponse] = None
    exercice: Optional[ExerciceBudgetaireResponse] = None
    created_by: Optional[UserSimpleResponse] = None
    projet_id: Optional[int] = None
    projet: Optional[ProjetSimpleResponse] = None


class BudgetDetailResponse(BudgetResponse):
    lignes: List[LigneBudgetaireSimpleResponse] = []

File: frontend/budget-managment/src/types/budget.ts
export type BudgetStatus =
  | "brouillon"
  | "soumis"
  | "soumis_gestionnaire"
  | "valide"
  | "valide_gestionnaire"
  | "soumis_admin"
  | "approuve_admin"
  | "rejete"
  | "rejete_gestionnaire"
  | "rejete_admin"
  | "en_execution"
  | "execute"
  | "cloture";

export interface Budget {
  id: number;
  reference: string;
  libelle: string;
  description?: string | null;
  devise?: "FC" | "USD";
  montant_total_prevu: number;
  montant_total_realise: number;
  total_recettes_realisees?: number;
  total_depenses_realisees?: number;
  taux_execution_budgetaire?: number;
  ecart_total: number;
  statut: BudgetStatus;
  date_creation?: string;
  departement_id: number;
  exercice_id: number;
  created_by_id?: number | null;
  projet_id?: number | null;
  exercice?: {
    id: number;
    libelle: string;
    date_debut: string;
    date_fin: string;
    statut: string;
  } | null;
  projet?: {
    id: number;
    code: string;
    titre: string;
    date_debut_prevue?: string | null;
    date_fin_prevue?: string | null;
    departement_id: number;
    chef_projet_id?: number | null;
    chef_projet?: {
      id: number;
      nom: string;
      prenom?: string | null;
      email: string;
    } | null;
    departement?: {
      id: number;
      nom: string;
    } | null;
  } | null;
}

export interface BudgetCreate {
  reference?: string;
  libelle: string;
  description?: string;
  devise?: "FC" | "USD";
  montant_total_prevu?: number;
  projet_id: number;
}

export interface BudgetUpdate {
  reference?: string;
  libelle?: string;
  description?: string;
  statut?: BudgetStatus;
}
